storage: filesystem/mmap, Add PackScanner to handle large repos

pjbgf opened the pull request:
The new PackScanner is able to navigate through a packfile and inflate objects on demand. This avoids costly operations such as loading into memory the entire index and resolving all delta objects which may or may not be used.

Performance-wise, the impact is negligible for small repositories, with the exception of the 60% drop in number of allocations:
```
BenchmarkPackHandlers/packfile-cache-osfs:_basic-fixture-16   	    4020	    288678 ns/op	  294714 B/op	     743 allocs/op
BenchmarkPackHandlers/packfile-nocache-memfs:basic-fixture-16 	    4272	    280586 ns/op	  294884 B/op	     742 allocs/op
BenchmarkPackHandlers/mmap-pack-scanner:basic-fixture-16      	    5646	    212349 ns/op	  293650 B/op	     286 allocs/op
```

In very large repositories is where most gains can be observed. For the linux kernel, the memory churn of fetching 4 objects drops from `1.3GB` to `1MB`.
Both allocations and time per operation are a fraction of the existing implementation:
```
BenchmarkPackHandlers/packfile-cache-osfs:_linux-kernel-16    	       1	2240495709 ns/op	1331433920 B/op	   35933 allocs/op
BenchmarkPackHandlers/packfile-nocache-memfs:linux-kernel-16  	       1	2026880001 ns/op	1331437040 B/op	   35935 allocs/op
BenchmarkPackHandlers/mmap-pack-scanner:linux-kernel-16       	     477	   2513275 ns/op	   1011512 B/op	    1488 allocs/op
```

Additionally, this PR:
- Refactors some internal pack functions so that it can be used by new and existing "Pack Handlers".
- Creates a new test/ dir, to amass tests that goes across implementations. This is also the home for some high-level benchmarks, which will help monitor performance improvements over time.


Relates to: https://github.com/go-git/go-git/issues/1451 https://github.com/go-git/go-git/issues/1691 https://github.com/go-git/go-git/issues/1673 https://github.com/go-git/go-git/issues/1601.
Part of: https://github.com/go-git/go-git/issues/1774.

This is part of a [openSUSE Hack Week 25 project](https://hackweek.opensuse.org/projects/go-git-unlocking-sha256-based-repository-cloning-ahead-of-git-v3).

Special thanks to @runxiyu for the initial nudge on mmap, which triggered the research that led to this PR. :bow:

pjbgf committed:
build: Bump go-billy and go-git-fixtures
The latest version of both dependencies enable the use of OSFixture
and Fd(), both being requirements to test the mmap implementation.

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

pjbgf committed:
plumbing: format/packfile, Extract pack and patch logic into new package

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

onee-only commented on storage/filesystem/mmap/fsobject.go:
```diff
@@ -0,0 +1,343 @@
+		base, err = o.scanner.GetByOffset(baseOffset)
+	} else {
+		base, err = o.scanner.Get(baseHash)
+	}
```
```suggestion
	// Get the base object to inherit its type.
	var base plumbing.EncodedObject
	var err error

	if o.diskType == plumbing.OFSDeltaObject {
		reader := bytes.NewReader(o.scanner.packMmap[pos:])
		negativeOffset, err := binary.ReadVariableWidthInt(reader)
		if err != nil {
			return fmt.Errorf("failed to read OFS delta offset: %w", err)
		}

		baseOffset := uint64(o.offset) - uint64(negativeOffset)
		consumed := len(o.scanner.packMmap[pos:]) - reader.Len()
		pos += int64(consumed)
		
		base, err = o.scanner.GetByOffset(baseOffset)
	} else {
		hashSize := o.scanner.hashSize
		if pos+int64(hashSize) > int64(len(o.scanner.packMmap)) {
			return fmt.Errorf("invalid REF delta: hash out of bounds")
		}

		baseHash, _ := plumbing.FromBytes(o.scanner.packMmap[pos : pos+int64(hashSize)])
		pos += int64(hashSize)
		
		base, err = o.scanner.Get(baseHash)
	}
```

IMO it looks cleaner to limit the scope of `baseOffset` and `baseHash`.

pjbgf replied:
Nice one, completely agree. :+1:

pjbgf commented on storage/filesystem/mmap/fsobject.go:
```diff
@@ -0,0 +1,343 @@
+		return io.NopCloser(br), nil
+	}
+
+	data := o.scanner.packMmap[start:]
```
That was my initial implementation, but when trying to limit the returning slice here I had some consistency issues - IIRC some objects would break the Zlib process. Note that this is already in place for the raw reader (`L95`) - which is slightly more important in this case.

Given that this is just a new slice, pointing to the same underlying array, and that the Zlib process stops at the end of the compressed data, I'm not sure it is worth changing it. But happy to change if you find a way around the issue I mentioned above.

pjbgf commented:
> There is one drawback to be aware of with mmap, you can end up with blocking the entire system thread on page fault while the OS fetch the missing file page. 

@cedric-appdirect Thanks for pointing that out. For context, handling very large repositories with the vanilla filesystem implementation results in GB's of memory churn. That combined with the impact of the GC pressure results in a very tough time for memory-constraint machines, as that may starve the rest of the system of both CPU and memory.

Now considering the benchmark I shared, the mmap implementation has a rough cost of `2.5ms` per operation. The page fault would need to make it at least a thousand times slower, for the performance to **match the current implementation**. 

> Or make the mmap code path optional.

This is already optional - as per PR description and linked issue.

cedric-appdirect commented:
I may have an interest in implementing the io.ReaderAt version as I use go-git in web service once this PR is merged. Let me know if you do not plan to work on it.

onee-only commented on storage/filesystem/mmap/files.go:
```diff
@@ -0,0 +1,90 @@
+)
+
+// mmapFile creates a memory-mapped region for the given file.
+func mmapFile(f billy.File) ([]byte, func() error, error) {
```
My thought was that the `billy.Filesystem` interface is conceptually based on an actual OS filesystem (especially posix). I’m aware that we probably wouldn’t benefit much from supporting mmap in filesystems like `memfs`, but I thought it should still be available for extensibility, similar to how we can mmap files on `tmpfs`.

I’m fine with either decision, though. It might be an unnecessary API addition for a very specific use case.

pjbgf committed:
storage: filesystem/mmap, Add PackScanner to handle large repos2
The new PackScanner is able to navigate through a packfile and inflate
objects on demand. This avoids costly operations such as loading into
memory the entire index and resolving all delta objects which may or
may not be used.

Performance-wise, the impact is negligible for small repositories,
with the exception of the 60% drop in number of allocations:

BenchmarkPackHandlers/packfile-cache-osfs:_basic-fixture-16   	    4020	    288678 ns/op	  294714 B/op	     743 allocs/op
BenchmarkPackHandlers/packfile-nocache-memfs:basic-fixture-16 	    4272	    280586 ns/op	  294884 B/op	     742 allocs/op
BenchmarkPackHandlers/mmap-pack-scanner:basic-fixture-16      	    5646	    212349 ns/op	  293650 B/op	     286 allocs/op

In very large repositories is where most gains can be observed. For the linux
kernel, the memory churn of fetching 4 objects drops from 1.3GB to 1MB.
Both allocations and time per operation are a fraction of the existing implementation:

BenchmarkPackHandlers/packfile-cache-osfs:_linux-kernel-16    	       1	2240495709 ns/op	1331433920 B/op	   35933 allocs/op
BenchmarkPackHandlers/packfile-nocache-memfs:linux-kernel-16  	       1	2026880001 ns/op	1331437040 B/op	   35935 allocs/op
BenchmarkPackHandlers/mmap-pack-scanner:linux-kernel-16       	     477	   2513275 ns/op	   1011512 B/op	    1488 allocs/op

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

pjbgf commented on tests/pack/common_test.go:
```diff
@@ -0,0 +1,66 @@
+	bufioPool = sync.Pool{New: func() any {
+		return bufio.NewReader(nil)
+	}}
+}
```
For future reference, while debugging this without `resetGlobalSyncPools`, `r` is `nil` as the pool is empty and the func `New` is nil and therefore not called (as stepped into golib code). As per stack trace, `init` is being called:

<img width="854" height="494" alt="image" src="https://github.com/user-attachments/assets/c9b525ba-4356-4777-bff3-0c5582129c47" />

```
--- FAIL: TestPackfileEmbedFS (0.00s)
    --- FAIL: TestPackfileEmbedFS/TestGet (0.00s)
        --- FAIL: TestPackfileEmbedFS/TestGet/valid_commit_hash (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7b4e18]

goroutine 29 [running]:
testing.tRunner.func1.2({0x89eaa0, 0x3b8c610})
	/usr/lib64/go/1.25/src/testing/testing.go:1872 +0x237
testing.tRunner.func1()
	/usr/lib64/go/1.25/src/testing/testing.go:1875 +0x35b
panic({0x89eaa0?, 0x3b8c610?})
	/usr/lib64/go/1.25/src/runtime/panic.go:783 +0x132
bufio.(*Reader).Reset(...)
	/usr/lib64/go/1.25/src/bufio/bufio.go:81
github.com/go-git/go-git/v6/utils/sync.GetBufioReader({0x0, 0x0})
	/home/levi/git/go-git/go-git/utils/sync/bufio.go:23 +0x78
github.com/go-git/go-git/v6/plumbing/format/packfile.(*Packfile).init.func1()
	/home/levi/git/go-git/go-git/plumbing/format/packfile/packfile.go:211 +0x39
sync.(*Once).doSlow(0x3b46440?, 0x8dd400?)
	/usr/lib64/go/1.25/src/sync/once.go:78 +0xac
sync.(*Once).Do(...)
	/usr/lib64/go/1.25/src/sync/once.go:69
github.com/go-git/go-git/v6/plumbing/format/packfile.(*Packfile).init(0xc0000c6370)
	/home/levi/git/go-git/go-git/plumbing/format/packfile/packfile.go:200 +0x45
github.com/go-git/go-git/v6/plumbing/format/packfile.(*Packfile).Get(0xc0000c6370, {{0xe8, 0xd3, 0xff, 0xab, 0x55, 0x28, 0x95, 0xc1, 0x9b, ...}, ...})
	/home/levi/git/go-git/go-git/plumbing/format/packfile/packfile.go:69 +0x37
github.com/go-git/go-git/v6/tests/pack_test.(*PackHandlerSuite[...]).TestGet.func1()
	/home/levi/git/go-git/go-git/tests/pack/pack_test.go:280 +0x52
testing.tRunner(0xc0001eb180, 0xc0000e0510)
	/usr/lib64/go/1.25/src/testing/testing.go:1934 +0xea
created by testing.(*T).Run in goroutine 28
	/usr/lib64/go/1.25/src/testing/testing.go:1997 +0x465
FAIL	github.com/go-git/go-git/v6/tests/pack	0.008s
FAIL
```

We can leave this as is for now, and review on a follow-up PR as this is related to the testing code as opposed to the main change proposed.

onee-only reviewed (approved):
LGTM. For some unknown reason, I can't resolve those reviews.

haines commented on storage/filesystem/mmap/fsobject.go:
```diff
@@ -0,0 +1,344 @@
+		pos += int64(consumed)
+
+		//nolint:staticcheck
+		base, err = o.scanner.GetByOffset(baseOffset) //nolint:ineffassign
```
This is actually correct because `err` is shadowed on `L179` (https://github.com/go-git/go-git/pull/2063).

runxiyu replied:
It's very rare for ineffassign to be wrong.
